src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java

```diff
@@ -567,79 +569,160 @@ public String getQuestionResultStatisticsHtml(List<FeedbackResponseAttributes> r
+     */
+    private List<FeedbackResponseAttributes> filterResponsesForStatistics(
+            List<FeedbackResponseAttributes> responses, FeedbackQuestionAttributes question,
+            String studentEmail, FeedbackSessionResultsBundle bundle, String view) {
+
+
+        boolean isViewedByStudent = "student".equals(view);
+        if (!isViewedByStudent) {
+            return responses;
+        }
+
+        FeedbackParticipantType recipientType = question.getRecipientType();
+
+        boolean isFilteringSkipped = recipientType.equals(FeedbackParticipantType.INSTRUCTORS)
+                || recipientType.equals(FeedbackParticipantType.NONE)
+                || recipientType.equals(FeedbackParticipantType.SELF);
+
+        if (isFilteringSkipped) {
+            return responses;
+        }
+
+        boolean isFilteringByTeams = recipientType.equals(FeedbackParticipantType.OWN_TEAM)
+                || recipientType.equals(FeedbackParticipantType.TEAMS);
+
+        List<FeedbackResponseAttributes> receivedResponses = new ArrayList<>();
+        String recipientString = isFilteringByTeams ? bundle.getTeamNameForEmail(studentEmail) : studentEmail;
+
+        for (FeedbackResponseAttributes response : responses) {
+            boolean isReceivedResponse = response.recipient.equals(recipientString);
+            if (isReceivedResponse) {
+                receivedResponses.add(response);
+            }
+        }
+
+        return receivedResponses;
+    }
     
     /**
      * Calculates the statistics for rubric question
      * 
      * Returns a 2D float array to indicate the percentage frequency
-     * a choice is selected for each sub-question.
+     * a choice is selected and the average weight for the responses for each sub-question.
+     *
+     * Values are set to 0 if there are no responses to that sub-question.
+     * Average value is set to 0 if there are no assigned weights.
      * 
      * e.g.
-     * pecentageFrequency[subQuestionIndex][choiceIndex]
+     * percentageFrequencyAndAverageValue[subQuestionIndex][choiceIndex]
      *  -> is the percentage choiceIndex is chosen for subQuestionIndex, for the given question/responses.
+     * percentageFrequencyAndAverageValue[subQuestionIndex][numOfRubricChoices]
+     *  -> is the average weight of the responses for the given sub-question
      *
      */
     private float[][] calculateRubricStats(List<FeedbackResponseAttributes> responses,
-            FeedbackQuestionAttributes question) {
-        FeedbackRubricQuestionDetails fqd = (FeedbackRubricQuestionDetails) question.getQuestionDetails();
-        
-        // Initialize response frequency variable, used to store frequency each choice is selected.
-        int[][] responseFrequency = calculateResponseFrequency(responses, fqd);
-        
-        return getPercentageFrequencyAndAverage(responseFrequency, fqd);
-    }
+                                           FeedbackRubricQuestionDetails feedbackQuestionDetails) {
```

Comment:
I modified the behaviour of the two methods and found the previous code quite hard to read so I thought it would be better to refactor it in this PR too.